e2e/testsuite/testsuite.go

```diff
@@ -611,6 +628,43 @@ func (s *E2ETestSuite) QueryGranterGrants(ctx context.Context, chain *cosmos.Cos
 	return grants.Grants, nil
 }
 
+// IsRelayer returns true if the provided wallet is used by a relayer.
+func (s *E2ETestSuite) IsRelayer(user ibc.Wallet) bool {
+	if s.relayers == nil {
+		return false
+	}
+	return s.relayers[user]
+}
+
+// retryNtimes retries the provided function up to the provided number of attempts.
+func (s *E2ETestSuite) retryNtimes(f func() (sdk.TxResponse, error), attempts int) (sdk.TxResponse, error) {
+	// Ignore account sequence mismatch errors.
+	allowedMessages := []string{"account sequence mismatch"}
+	var err error
+	var resp sdk.TxResponse
+	for i := 0; i < attempts; i++ {
+		// If the response's raw log doesn't contain any of the allowed prefixes we return, else, we retry.
+		resp, err = f()
```

Comment:
Small optimization: should we also check that `err` is not nil? Otherwise, if it is `nil` then the transaction should be succeeded, right? And we don't need to check the response log or retry.

Reply:
As far as I've understood the issue, `err` is `nil` when this happens. The test was failing when we tried to check that the `resp` was a valid tx response [here](https://github.com/cosmos/ibc-go/blob/d840c699d988e32cb55f2c289d69f8ef2e5cea50/e2e/testsuite/testsuite.go#L431) and it failed when `gasUsed == 0`.

Happy to change if I misunderstood something.

Reply:
does it make sense to add an additional check for the error here? Should we retry the message if an error occurred but `RawLog` contains one of the retry messages?

Reply:
hm, maybe we can do

```go
resp, err = f()

if !s.IsValidTxResponse(resp) {
    continue
}

if !containsMessage(resp.RawLog, retryMessages) {
    return resp, err
}
// continue loop 
```

This would involve adding a function `IsValidTxResponse` (which would just do what the assertion function does but return a bool I guess)

Reply:
I'm probably missing something here :smile: I was under the impression we'd like to skip invalid tx's that contain those messages in their `RawLog`, any other invalid `tx`s are just returned back to broadcast message which will eventually pass them over to its callee and have `AssertValidTxResponse` fail on us.

If we add a call to check if its a valid response beforehand that would `continue` without leaving us the chance to check for the `RawLog` (basically a retry again but for cases where we might not want that). 

I do think I'm not seeing something obvious though.

Reply:
you are definitely correct! I was getting mixed up!   We definitely would need to do the contains check first. 

Thinking about it some more, I think we may be able to fail in the case of an error, and exit entirely then, but retry if the raw logs contain one of these strings.

The failures we're seeing because of the sequence mismatch are returning a nil error, so an error in the broadcast itself can still be considered a failure.

Reply:
I've added the check for err as suggested. I'll open the issue if there's at least a 2/3 consensus.

Reply:
I can open an issue and we can discuss it on there